p2p/src/authenticated/discovery/actors/listener.rs

```diff
@@ -91,13 +180,24 @@ impl<E: Spawner + Clock + ReasonablyRealtime + Network + Rng + CryptoRng + Metri
         };
         debug!(?peer, ?address, "completed handshake");
 
+        if peer_rate_limiter.check_key(&peer).is_err() {
+            handshakes_peer_rate_limited.inc();
+            debug!(?peer, ?address, "peer exceeded handshake rate limit");
+            return;
+        }
+
         // Attempt to claim the connection
         let Some(reservation) = tracker.listen(peer.clone()).await else {
             debug!(?peer, ?address, "unable to reserve connection to peer");
             return;
         };
         debug!(?peer, ?address, "reserved connection");
 
+        {
+            let mut guard = prioritized_ips.write().await;
+            guard.entry(address.ip()).or_default().insert(peer.clone());
```

Comment:
If some peer connects on a different IP, we should drop the old one (otherwise it is trivial to OOM this map)